include/seqan3/io/alignment_file/detail.hpp

```diff
@@ -153,12 +157,15 @@ char compare_aligned_values(reference_char_type const reference_char,
  * The following cigar string when printed: "4M2I5M2D1M". The extended cigar
  * string would look like this: "3=1X2I3=1X1=2D1=".
  */
-template<std::ranges::ForwardRange ref_seq_type, std::ranges::ForwardRange query_seq_type>
-//\!cond
-    requires std::EqualityComparableWith<gap, value_type_t<ref_seq_type>> &&
-             std::EqualityComparableWith<gap, value_type_t<query_seq_type>>
-//\!endcond
-std::string get_cigar_string(std::pair<ref_seq_type, query_seq_type> const & alignment,
+template<tuple_like_concept alignment_type>
+//!\cond
+    requires requires {{std::tuple_size_v<alignment_type> == 2}} &&
+             std::ranges::ForwardRange<remove_cvref_t<decltype(get<0>(alignment_type{}))>> &&
+             std::ranges::ForwardRange<remove_cvref_t<decltype(get<1>(alignment_type{}))>> &&
+             std::EqualityComparableWith<gap, value_type_t<remove_cvref_t<decltype(get<0>(alignment_type{}))>>> &&
+             std::EqualityComparableWith<gap, value_type_t<remove_cvref_t<decltype(get<1>(alignment_type{}))>>>
```

Comment:
1) why do we need remove_cvref_t here?
2) use `std::tuple_element_t<>` instead of `decltype(get<`
3) use `alignment_type && alignment` instead of `const &`

But in general I think this whole signature could be greatly simplified by just taking two range arguments as `&&`. That would make the concept checks much easier to write (and to explain in the documentation) and also because they are named it would be immediately clear which one is the reference and which one the query.

Reply:
1) I thought that's what you meant in your comment "get_string constraints are not guarded". And I thought value_type does not always work on references.

2) why removing `const`?

general) well, but we decided to treat "an alignment" as a pair of gapped sequences. so I thought we at least make the related functions work on "an alignment".

Reply:
> I thought that's what you meant in your comment "get_string constraints are not guarded"

no, I meant that `//!\cond` is missing around it :laughing: 

> And I thought value_type does not always work on references.

It should definitely work on ranges, otherwise that's a bug.

> why removing const?

you mean why remove `remove_cvref_t`? See my answer below, I think it's not only unnecessary, but actually wrong.

In general, please don't just `remove_cvref_t` everywhere that you are not sure, but actually try to understand what's going on :wink: In most cases you won't need it and in many others `remove_const_t` or `remove_reference_t` will be sufficient.

> general) well, but we decided to treat "an alignment" as a pair of gapped sequences. so I thought we at least make the related functions work on "an alignment".

we agreed internally that we wouldn't make an extra data type but that doesn't mean we have to use pairs or tuples everywhere. I think that depends on the situation, e.g if you compare two different "alignments", then I would take two tuples so that each row is clearly grouped into one of the two, but in this case the interface would just become simpler by individual args IMHO.
Especially since both rows of the alignment are not treated equally by this algorithm!

Reply:
> you mean why remove remove_cvref_t?

No sorry that was meant for point 3) why passing `alignment_type && alignment` and not `alignment_type const && alignment`

Reply:
because the tuple could contain ranges that are not const-iterable and if the tuple is const the elements will appear const, too, and it will break.

It's best to test all these functions that take "some range" with a range that is not const-iterable, like `view::drop`.

Reply:
(but this would also disappear if the tuple disappears!

Reply:
Just to add to it. The tuple interface makes the typical use case working with seqan alignment computation and much more easier, as it does not involve untying tuple before passing it's elements into the function and fairly I was hoping to have this interface at the end:
```cpp
debug_stream << aling_pairwise(...) | transform([](auto && res){ return res.trace(); }) | to_cigar << '\n';
```